develop/tutorials/articles-dxp/310-troubleshooting/13-patching-and-debugging-portal-source-code.markdown

```diff
@@ -0,0 +1,192 @@
+    `[LIFERAY_HOME]/patching-tool/patches`.
+
+2.  Open a terminal to the folder `[LIFERAY_HOME]/patching-tool`.
+
+3.  Generate a default patching profile called `default.properties` by running
+    `patching-tool.sh auto-discovery`. Make sure the profile's properties
+    refers to your DXP installation. See
+    [the patching tool documentation](/discover/deployment/-/knowledge_base/7-0/patching-tool)
+    for more details.
+
+    Here's an example profile:
+
+    	patching.mode=binary
+    	war.path=../tomcat-8.0.32/webapps/ROOT/
+    	global.lib.path=../tomcat-8.0.32/lib/ext/
+    	liferay.home=../
+
+4.  Execute the command `patching-tool.sh info` to list your installed
+    patches and patches available in your `patches` folder.
+
+5.  Install all available local patches by executing this command:
+
+        patching-tool.sh install
+
+    The
+    [patching tool documentation](/discover/deployment/-/knowledge_base/7-0/patching-tool)
+    explains additional steps that might apply, such as deleting your
+    `[LIFERAY_HOME]/osgi/state` folder and creating database indexes.
+
+You're ready to prepare the DXP source code for debugging. 
+
+## Step 3: Prepare the DXP Source Code [](id=prepare-the-dxp-source-code)
+
+Version control systems, such as [Git](https://git-scm.com/) facilitate
+examining code changes. To help you compare patches, it's recommended to commit
+the DXP source code and each patch separately to a version control system.
+Seeing the diffs between the patch commits can help you find where behaviors are
+introduced. This tutorial uses the Git version control system. 
+
+Here are steps for preparing the source:
+
+1.  [Commit the source code to version control.](#commit-the-dxp-source-code)
+2.  [Create a patching tool profile for the source code.](#create-a-patching-tool-profile-for-the-source-code)
+3.  [Patch the source.](patch-the-source-code)
+
+### Commit the DXP Source Code [](id=commit-the-dxp-source-code)
+
+The DXP source code is your initial code base. You'll commit it to version
+control before patching it. 
+
+1.  Unzip the DXP source code to a folder where you want to work with it.
+
+2.  In a terminal, commit the DXP source code to a Git repository by executing
+    these commands:
+
+        cd [path to source code root folder]
+        git init
+        git add .
+        git commit -a
+
+Here's are the command descriptions:
+
+-   `init` creates a Git repository for the current folder (i.e., the root
+    folder) and all its contents.
+-   `add` stages the root folder and its contents.
+-   `commit` checks in the staged files.
+
+### Create a Patching Tool Profile for the Source Code [](id=create-a-patching-tool-profile-for-the-source-code)
+
+Next, you need a patching tool profile that points to your source code.  
+
+1.  Execute the following command to create a profile. Replace `[profile]`
+    with an arbitrary name for your profile. 
+
+        patching-tool.sh auto-discovery [profile]
```

Comment:
If you call `patching-too.sh src auto-discovery path-to-source source` you'll get the configuration prepared in the src folder without editing.

Reply:
Wow. Very cool.